src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -1291,13 +1286,35 @@ private BoundExpression PopulateDictionary(BoundCollectionExpression node, Named
                                 rewrittenExpression,
                                 iteratorBody =>
                                 {
-                                    // dictionary[item.Key] = item.Value;
-                                    var expr = VisitExpression(((BoundExpressionStatement)iteratorBody).Expression);
-                                    return new BoundExpressionStatement(expr.Syntax, expr);
+                                    var localsBuilder = ArrayBuilder<BoundLocal>.GetInstance();
+                                    var sideEffects = ArrayBuilder<BoundExpression>.GetInstance();
+                                    var expr = ((BoundExpressionStatement)iteratorBody).Expression;
+                                    switch (expr)
+                                    {
+                                        case BoundKeyValuePairConversion keyValuePairConversion:
+                                            // dictionary[(TKey)item.Key] = (TValue)item.Value;
+                                            assignItemFromKeyValuePairConversion(keyValuePairConversion, dictionaryTemp, setMethod, localsBuilder, sideEffects);
+                                            break;
+                                        case BoundExpression expressionElement:
+                                            // dictionary[element.Key] = element.Value;
+                                            assignItemFromExpression(expressionElement, dictionaryTemp, setMethod, localsBuilder, sideEffects);
+                                            break;
+                                        default:
+                                            throw ExceptionUtilities.UnexpectedValue(element);
+                                    }
+                                    var locals = localsBuilder.SelectAsArray(local => local.LocalSymbol);
+                                    var statements = sideEffects.SelectAsArray(expr => (BoundStatement)new BoundExpressionStatement(expr.Syntax, expr));
+                                    sideEffects.Free();
+                                    localsBuilder.Free();
+                                    return new BoundBlock(iteratorBody.Syntax, locals, statements);
                                 });
                             sideEffects.Add(rewrittenElement);
                         }
                         break;
+                    case BoundExpression expressionElement:
```

Comment:
nit: consider moving this case under the `BoundKeyValuePairConversion`. That way the parallel between those two top-level cases and the two cases within the `BoundCollectionExpressionSpreadElement` is more evident